File: pipeline/plan/super_series.py
"""Super Series — super-plan validation and slicing (Python side).

Paired 1:1 with ``bot/src/super_series.js`` — both modules implement the exact
same accept/reject decisions and error strings for the super-plan contract
(documented in ARCHITECTURE.md §7.5). Cross-side equivalence is enforced by
``pipeline/tests/test_super_series.py`` driving the JS side through
``pipeline/tests/run_js_validator.mjs``-style node invocation.

DESIGN (feature-01)
-------------------

A Super Series task's Stage A prompt asks the external AI to plan the ENTIRE
series in ONE document (the "super-plan"): a ``parts`` array of ordinary,
self-contained per-part production.json documents (ARCHITECTURE.md §7.3
shape), each already carrying its own correct ``series`` block, plus a shared
``series_id``. The bot validates the whole document up front (this module's
``validate_super_plan``) BEFORE slicing anything, then hands out exactly one
part at a time to the completely unmodified Stage B pipeline — from Stage B's
perspective every Super Series part is an ordinary series part.

Validation rigor deliberately matches ``schema.validate_production_plan``:
every part is run through that exact validator with a ``part_number``
override, then cross-part checks (contiguous tiling, one is_final, distinct
titles, consistent series_id) are applied on top.
"""

from __future__ import annotations

import json
from typing import Any

from pipeline.plan import schema as plan_schema

MAX_PARTS = 20


def _is_int(value: Any) -> bool:
    return plan_schema._is_int(value)


def _as_int(value: Any) -> int:
    return plan_schema._as_int(value)


def _is_nonempty_string(value: Any) -> bool:
    return plan_schema._is_nonempty_string(value)


def validate_super_plan(document: Any) -> list[str]:
    """Return a list of validation error strings for a super-plan document.

    An empty list means the document is valid and safe to slice. Errors are
    written so an external AI (or operator) can fix the document and re-send
    it; each message names the exact field and part index involved.
    """
    errors: list[str] = []

    if not isinstance(document, dict):
        return ["Top level must be a JSON object."]

    # -- Shared series id --------------------------------------------------- #
    series_id = document.get("series_id")
    if not _is_nonempty_string(series_id):
        errors.append("`series_id` must be a non-empty string shared by every part.")
        series_id = None

    # -- Required positive-integer scalars ---------------------------------- #
    video_duration = document.get("video_duration_seconds")
    if not _is_int(video_duration) or _as_int(video_duration) <= 0:
        errors.append("`video_duration_seconds` must be a positive integer.")
        video_duration = None
    target_duration = document.get("target_total_duration_seconds")
    if not _is_int(target_duration) or _as_int(target_duration) <= 0:
        errors.append("`target_total_duration_seconds` must be a positive integer.")

    # -- parts array --------------------------------------------------------- #
    parts = document.get("parts")
    if not isinstance(parts, list):
        errors.append("`parts` must be an array of per-part production plans.")
        return errors
    if len(parts) < 1:
        errors.append("`parts` is empty — at least one part is required.")
        return errors
    if len(parts) > MAX_PARTS:
        errors.append(f"`parts` must contain at most {MAX_PARTS} entries.")

    final_count = 0
    titles: set[str] = set()
    previous_end: int | None = None

    for index, part in enumerate(parts):
        at = f"parts[{index}]"
        if not isinstance(part, dict):
            errors.append(f"{at} must be an object.")
            continue

        # Every part must carry the same shared series_id.
        part_series = part.get("series") if isinstance(part.get("series"), dict) else {}
        part_sid = part_series.get("series_id")
        if series_id is not None and part_sid != series_id:
            errors.append(f"{at}.series.series_id must equal the shared top-level series_id.")

        # Each part is itself an ordinary §7.3 series production plan; run the
        # existing single-part validator against it with the part number
        # derived from its position (the super-plan's parts are positional).
        part_errors = plan_schema.validate_production_plan(part, part_number=index + 1)
        errors.extend(f"{at}: {msg}" for msg in part_errors)

        # Cross-part checks (only meaningful when the part-level fields parsed).
        part_final = part_series.get("is_final")
        if part_final is True:
            final_count += 1

        title = part.get("title")
        if _is_nonempty_string(title):
            key = title.strip().lower()
            if key in titles:
                errors.append(f"{at}.title duplicates an earlier part's title.")
            else:
                titles.add(key)

        start_val = part_series.get("start_seconds")
        end_val = part_series.get("end_seconds")
        if _is_int(start_val) and _is_int(end_val):
            start = _as_int(start_val)
            end = _as_int(end_val)
            if index == 0 and start != 0:
                errors.append(f"{at}.series.start_seconds must be 0 for the first part.")
            if previous_end is not None and start != previous_end:
                errors.append(
                    f"{at}.series.start_seconds must equal the previous part's series_end_seconds "
                    f"(parts must tile the source with no gaps or overlaps)."
                )
            previous_end = end

    if isinstance(parts, list) and parts:
        if final_count != 1:
            errors.append("Exactly one part must be marked series.is_final = true.")
        else:
            # The single final part must be the LAST one — a final marker in
            # the middle would strand every later part unqueued.
            last_series = parts[-1].get("series") if isinstance(parts[-1], dict) and isinstance(parts[-1].get("series"), dict) else {}
            if last_series.get("is_final") is not True:
                errors.append("The part marked series.is_final = true must be the last entry in `parts`.")

    return errors


def slice_part(document: dict[str, Any], part_index: int) -> dict[str, Any]:
    """Return the ordinary single-part production.json for ``part_index`` (0-based).

    Callers MUST run ``validate_super_plan`` first and only call this on a
    valid document. The returned document is a plain §7.3 production plan —
    identical in shape to what a normal Series Mode part's AI would return —
    with its ``series.part`` field re-stamped to the positional part number.
    """
    parts = document["parts"]
    part = dict(parts[part_index])
    series = dict(part.get("series") if isinstance(part.get("series"), dict) else {})
    series["part"] = part_index + 1
    part["series"] = series
    part["job_id"] = ""  # stamped by the caller for the real spawned job id
    return part


def parse_and_validate_super_plan(text: str) -> tuple[dict[str, Any] | None, list[str]]:
    """Parse JSON ``text`` and validate it as a super-plan.

    Returns ``(document_or_None, errors)``. On JSON parse failure the document
    is ``None`` and errors contains a single "Not valid JSON: …" line.
    """
    try:
        document = json.loads(text)
    except json.JSONDecodeError as exc:
        return None, [f"Not valid JSON: {exc.msg}."]
    return document, validate_super_plan(document)


__all__ = [
    "MAX_PARTS",
    "validate_super_plan",
    "slice_part",
    "parse_and_validate_super_plan",
]

File: bot/src/super_series.js
/**
 * Super Series — super-plan validation, slicing, and the sequential queue
 * controller (Bot A side). Paired 1:1 with ``pipeline/plan/super_series.py``:
 * the validators on both sides must return the same accept/reject decisions
 * and the same error strings for the same input (pinned by
 * pipeline/tests/test_super_series.py).
 *
 * DESIGN (feature-01)
 * -------------------
 * The operator pastes ONE super-plan document (see ARCHITECTURE.md §7.5):
 * a ``parts`` array of ordinary, self-contained per-part production.json
 * documents plus a shared ``series_id``. This module:
 *
 *   1. validates the WHOLE document up front (validateSuperPlan) — malformed
 *      documents, non-tiling parts, or anything other than exactly one
 *      is_final reject the entire plan before any render is queued;
 *   2. slices out exactly one part at a time (sliceSuperPart) into an
 *      ordinary §7.3 single-part production.json and a §7.1 stage-a-request
 *      body synthesized with the EXISTING series continuation helpers
 *      (bot/src/series.js nextPartRequestBody / nextPartJobId) — the spawned
 *      job is indistinguishable from a normal series part;
 *   3. decides the queue's next move from durable repo state alone
 *      (superQueueAdvance): the controller is a pure decision function called
 *      from the bot's existing per-minute cron sweep, so it survives worker
 *      restarts and never double-dispatches.
 *
 * HALT-AND-RESUME (operator requirement): when a part's Stage B run reaches
 * state "error", the queue HALTS — nothing further is queued and the operator
 * is told to fix/restart that part through the existing task:restartb flow.
 * When that part later reaches "complete", the same decision function simply
 * returns the next part to queue — resumption needs no separate trigger.
 */

import { isValidJobId } from './jobs.js';
import { validateProductionPlan } from './plan.js';
import { buildSeriesContext, nextPartJobId, nextPartRequestBody } from './series.js';

export const MAX_PARTS = 20;
export const SUPER_STATE_PREFIX = 'super-state:';

function isPlainObject(value) {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function isInteger(value) {
  return typeof value === 'number' && Number.isFinite(value) && Math.floor(value) === value;
}

function isNonemptyString(value) {
  return typeof value === 'string' && value.trim() !== '';
}

// --------------------------------------------------------------------------- //
// Validation                                                                   //
// --------------------------------------------------------------------------- //

/**
 * Validate a super-plan document. Returns an array of error strings; empty
 * means valid and safe to slice. Mirrors pipeline/plan/super_series.py's
 * validate_super_plan exactly.
 */
export function validateSuperPlan(document) {
  const errors = [];

  if (!isPlainObject(document)) {
    return ['Top level must be a JSON object.'];
  }

  // -- Shared series id ---------------------------------------------------- //
  let seriesId = document.series_id;
  if (!isNonemptyString(seriesId)) {
    errors.push('`series_id` must be a non-empty string shared by every part.');
    seriesId = null;
  }

  // -- Required positive-integer scalars ----------------------------------- //
  const videoDuration = document.video_duration_seconds;
  if (!isInteger(videoDuration) || videoDuration <= 0) {
    errors.push('`video_duration_seconds` must be a positive integer.');
  }
  const targetDuration = document.target_total_duration_seconds;
  if (!isInteger(targetDuration) || targetDuration <= 0) {
    errors.push('`target_total_duration_seconds` must be a positive integer.');
  }

  // -- parts array ---------------------------------------------------------- //
  const parts = document.parts;
  if (!Array.isArray(parts)) {
    errors.push('`parts` must be an array of per-part production plans.');
    return errors;
  }
  if (parts.length < 1) {
    errors.push('`parts` is empty — at least one part is required.');
    return errors;
  }
  if (parts.length > MAX_PARTS) {
    errors.push(`\`parts\` must contain at most ${MAX_PARTS} entries.`);
  }

  let finalCount = 0;
  const titles = new Set();
  let previousEnd = null;

  for (let index = 0; index < parts.length; index += 1) {
    const at = `parts[${index}]`;
    const part = parts[index];
    if (!isPlainObject(part)) {
      errors.push(`${at} must be an object.`);
      continue;
    }

    // Every part must carry the same shared series_id.
    const partSeries = isPlainObject(part.series) ? part.series : {};
    if (seriesId !== null && partSeries.series_id !== seriesId) {
      errors.push(`${at}.series.series_id must equal the shared top-level series_id.`);
    }

    // Each part is itself an ordinary §7.3 series production plan; run the
    // existing single-part validator with the positional part number.
    const partErrors = validateProductionPlan(part, { partNumber: index + 1 });
    for (const message of partErrors) errors.push(`${at}: ${message}`);

    if (partSeries.is_final === true) finalCount += 1;

    if (isNonemptyString(part.title)) {
      const key = part.title.trim().toLowerCase();
      if (titles.has(key)) {
        errors.push(`${at}.title duplicates an earlier part's title.`);
      } else {
        titles.add(key);
      }
    }

    const startVal = partSeries.start_seconds;
    const endVal = partSeries.end_seconds;
    if (isInteger(startVal) && isInteger(endVal)) {
      if (index === 0 && startVal !== 0) {
        errors.push(`${at}.series.start_seconds must be 0 for the first part.`);
      }
      if (previousEnd !== null && startVal !== previousEnd) {
        errors.push(
          `${at}.series.start_seconds must equal the previous part's series_end_seconds ` +
          '(parts must tile the source with no gaps or overlaps).'
        );
      }
      previousEnd = endVal;
    }
  }

  if (finalCount !== 1) {
    errors.push('Exactly one part must be marked series.is_final = true.');
  } else {
    const last = parts[parts.length - 1];
    const lastSeries = isPlainObject(last) && isPlainObject(last.series) ? last.series : {};
    if (lastSeries.is_final !== true) {
      errors.push('The part marked series.is_final = true must be the last entry in `parts`.');
    }
  }

  return errors;
}

/**
 * Parse JSON text and validate it as a super-plan. Mirrors
 * pipeline/plan/super_series.py's parse_and_validate_super_plan.
 */
export function parseAndValidateSuperPlan(text) {
  let document;
  try {
    document = JSON.parse(text);
  } catch (error) {
    return { document: null, errors: [`Not valid JSON: ${error && error.message ? error.message : 'parse error'}.`] };
  }
  return { document, errors: validateSuperPlan(document) };
}

// --------------------------------------------------------------------------- //
// Slicing                                                                      //
// --------------------------------------------------------------------------- //

/**
 * Return the ordinary single-part production.json for ``partIndex`` (0-based)
 * of a VALID super-plan: the part's own document, with ``series.part``
 * re-stamped to its positional number. ``jobId`` is the real spawned job id.
 */
export function sliceSuperPart(document, partIndex, jobId) {
  const source = document.parts[partIndex];
  const part = JSON.parse(JSON.stringify(source));
  const series = isPlainObject(part.series) ? part.series : {};
  series.part = partIndex + 1;
  part.series = series;
  part.job_id = String(jobId || '');
  return part;
}

// --------------------------------------------------------------------------- //
// Durable queue state (stored at jobs/<anchor>/super-plan.json)              //
// --------------------------------------------------------------------------- //

/**
 * The durable record the queue controller re-reads on every sweep tick. It
 * lives in the anchor job's directory (the Super Series Stage A job) so the
 * existing cleanup/TTL machinery expires it together with the job, and so a
 * worker restart loses nothing.
 */
export function buildSuperState({ anchorJobId, seriesId, document, spawned = [] }) {
  return {
    version: 1,
    anchor_job_id: String(anchorJobId),
    series_id: String(seriesId),
    total_parts: document.parts.length,
    video_duration_seconds: Number(document.video_duration_seconds),
    // The whole validated super-plan travels with the record — every later
    // slice needs it, and re-reading one small JSON per tick is cheap.
    plan: document,
    // [{ part, job_id }] in spawn order; the controller derives "where am I"
    // from this plus each spawned job's status.json — no separate cursor.
    spawned: spawned.map((entry) => ({ part: Number(entry.part), job_id: String(entry.job_id) })),
  };
}

/**
 * Pure queue-decision function. Given the durable super state and a lookup
 * for spawned jobs' current status.json state, decide the next action.
 *
 * Returns one of:
 *   { action: 'queue',  part, jobId, plan }   — slice + dispatch this part now
 *   { action: 'halted', part, jobId, message } — a spawned part is in error;
 *                                                operator must restart it
 *   { action: 'waiting', part, jobId }         — a spawned part is still running
 *   { action: 'done' }                          — every part completed
 *
 * HALT-AND-RESUME is expressed entirely by the ordering below: the FIRST
 * spawned part that is not yet "complete" decides everything. Error → halt.
 * Anything else non-terminal → wait. Complete → look at the next part, and if
 * none remain, done. A manually-restarted part that reaches "complete"
 * therefore resumes the queue automatically on the next sweep tick.
 */
export function superQueueAdvance(state, statusFor) {
  const plan = state && state.plan;
  const totalParts = Number(state && state.total_parts) || 0;
  if (!isPlainObject(plan) || !Array.isArray(plan.parts) || totalParts < 1) {
    return { action: 'done' };
  }
  const spawned = Array.isArray(state.spawned) ? state.spawned : [];

  // 1. Walk spawned parts in order; the first not-complete part decides.
  for (const entry of spawned) {
    const jobState = String((statusFor(entry.job_id) || {}).state || '');
    if (jobState === 'complete') continue;
    if (jobState === 'error') {
      return {
        action: 'halted',
        part: Number(entry.part),
        jobId: entry.job_id,
        message:
          `Super Series part ${entry.part} of ${totalParts} failed (task ${entry.job_id} is in state error). ` +
          'The queue is PAUSED — no further parts will be queued. Open that task and use Restart Stage B; ' +
          'the moment it reaches complete, the queue resumes automatically.',
      };
    }
    if (jobState === 'cancelled') {
      return {
        action: 'halted',
        part: Number(entry.part),
        jobId: entry.job_id,
        message:
          `Super Series part ${entry.part} of ${totalParts} was cancelled (task ${entry.job_id}). ` +
          'The queue is PAUSED — no further parts will be queued. Open that task and use Restart Stage B; ' +
          'the moment it reaches complete, the queue resumes automatically.',
      };
    }
    // queued / stage_b_queued / stage_b_running / anything else non-terminal:
    // the pipeline is working on it — the queue waits for it.
    return { action: 'waiting', part: Number(entry.part), jobId: entry.job_id };
  }

  // 2. Everything spawned so far is complete.
  if (spawned.length >= totalParts) return { action: 'done' };

  // 3. Slice and queue the next part.
  const partNumber = spawned.length + 1;
  const jobId = superPartJobId(state.series_id, partNumber);
  return {
    action: 'queue',
    part: partNumber,
    jobId,
    plan: sliceSuperPart(plan, partNumber - 1, jobId),
  };
}

/** §6.3-safe job id for a spawned Super Series part (same rule as series.js). */
export function superPartJobId(seriesId, partNumber) {
  return nextPartJobId({ seriesId, part: partNumber });
}

/**
 * Synthesize the ordinary §7.1 stage-a-request body for a spawned Super
 * Series part, reusing the EXISTING series continuation helper
 * (series.js nextPartRequestBody) so the result is byte-for-byte what a
 * normal manual continuation would have written. ``anchorRequest`` is the
 * Super Series Stage A job's own request; ``summaries`` is the array of
 * { part, summary } entries from already-completed spawned parts (the same
 * continuity context a normal series part would carry).
 */
export function superPartRequestBody(anchorRequest, state, partNumber, summaries) {
  const context = buildSeriesContext(summaries);
  return nextPartRequestBody(
    anchorRequest,
    { seriesId: state.series_id, part: partNumber, startSeconds: Number(planStartSeconds(state, partNumber)) },
    context,
    String(anchorRequest && anchorRequest.job_id || ''),
  );
}

function planStartSeconds(state, partNumber) {
  const part = state && state.plan && state.plan.parts ? state.plan.parts[partNumber - 1] : null;
  const series = isPlainObject(part) && isPlainObject(part.series) ? part.series : {};
  return isInteger(series.start_seconds) ? series.start_seconds : 0;
}

export { isValidJobId };
